server/src/main/java/org/elasticsearch/discovery/HandshakingTransportAddressConnector.java

```diff
@@ -99,17 +100,23 @@ protected void innerOnResponse(DiscoveryNode remoteNode) {
                                     IOUtils.closeWhileHandlingException(connection);
 
                                     if (remoteNode.equals(transportService.getLocalNode())) {
-                                        // TODO cache this result for some time? forever?
                                         listener.onFailure(new ConnectTransportException(remoteNode, "local node found"));
                                     } else if (remoteNode.isMasterNode() == false) {
-                                        // TODO cache this result for some time?
                                         listener.onFailure(new ConnectTransportException(remoteNode, "non-master-eligible node found"));
                                     } else {
-                                        transportService.connectToNode(remoteNode, ActionListener.delegateFailure(listener,
-                                            (l, ignored) -> {
-                                                logger.trace("[{}] full connection successful: {}", thisConnectionAttempt, remoteNode);
-                                                listener.onResponse(remoteNode);
-                                            }));
+                                        transportService.connectToNode(remoteNode, ActionListener.wrap(ignored -> {
```

Comment:
Why move to `wrap` here? We (mostly Henning :)) are currently trying to remove the number of instances of passing broken listeners to transport APIs that don't handle their own exceptions and this seems like a step in the wrong direction. Can we fix the listener to handle its exception instead?